File: hard/min_path_sum.rb
n = i = 0
b = []
File.open(ARGV[0]).each_line do |line|
  if n == 0
    n = line.to_i
    i = n
    next
  end
  s = line.split(',')
  if i == n
    b = Array.new(n)
    b[0] = s[0].to_i
    (1...n).each do |j|
      b[j] = b[j - 1] + s[j].to_i
    end
  else
    b[0] += s[0].to_i
    (1...n).each do |j|
      b[j] = [b[j - 1], b[j]].min + s[j].to_i
    end
  end
  i -= 1
  if i == 0
    puts b[-1]
    n = 0
  end
end

File: hard/min_path_sum.pl
use strict;
use List::Util qw[min];
my ( $n, $i ) = ( 0, 0 );
my @b;
open( INFILE, "<", $ARGV[0] )
  or die("Cannot open file $ARGV[0] for reading: $!");
while ( my $line = <INFILE> ) {
    chomp $line;
    if ( $n == 0 ) {
        ( $n, $i ) = ( $line, $line );
        next;
    }
    my @s = split( /,/, $line );
    if ( $i == $n ) {
        @b = ();
        $b[0] = $s[0];
        for ( my $j = 1 ; $j < $n ; $j++ ) {
            $b[$j] = $b[ $j - 1 ] + $s[$j];
        }
    }
    else {
        $b[0] += $s[0];
        for ( my $j = 1 ; $j < $n ; $j++ ) {
            $b[$j] = min( $b[ $j - 1 ], $b[$j] ) + $s[$j];
        }
    }
    $i--;
    if ( $i == 0 ) {
        print "$b[-1]\n";
        $n = 0;
    }
}
close(INFILE);
